### ISSUE TITLE
catalogrule_group_website always refilling while indexing and a criticism about catalogrule_product indexer

### ISSUE THREAD

**Issue by cizgidekiadam** [magento/magento2#9906]
I figured that while waiting a looonng index processing to complete (It is not to be completed for a couple days). 

While trying to track tables changes (it is funny) desperately, i see that catalogrule_group_website table refilling continuously. 

I thought it is strange so look at `applyAllRules` method in 'module-catalog-rule/Model/Indexer/IndexBuilder', there is a call to `updateCatalogRuleGroupWebsiteData`.

It should call `updateCatalogRuleGroupWebsiteData` method at the end of indexing process, because it just get website data from `catalog_rule_product' table by distinct, so website data is not changing while indexer working.

With current way, there is unnecessary database queries (event little bit) again and again (it is so annoying). 

Another thing, we have ~2.7 million records in `catalogrule_product`, and ~7.5 million records in `catalogrule_product_price` table.

When Magento needs to be reindex (even not required a full index) it is deleting all data in `catalogrule_product` table for each rule and refilling again. As a result, indexing process taking too long and while it is happening, at frontend, discounts not working/appearing correctly. 

I didn't understood that logic.
Is there an explanation for it ?

### Preconditions

1. Magento 2.1.2
2. PHP 7.0.18
3. Nginx with FPM support
4. Varnish Cache
5. Redis Session Cache
6. A bold coffee

### Steps to reproduce

1. Watch database tables while indexing :)

### Expected result

1. Nothing expected.

### Actual result

1. Thousands/millions of database queries (which is not using query cache, because it it inserting)
2. Tired eyes by trying to track database changes :'(

**Comment by magento-engcom-team** [comment]
@rbostan, thank you for your report.
We've created internal ticket(s) MAGETWO-84359 to track progress on the issue.

**Comment by SimonPrins** [comment]
I have the same issue (Magento 2.3.2). I have about 130000 products in the shop, with 6 rules. Updating the indexer has taken hours so far and still is not complete. I constantly see queries to fill the `catalogrule_group_website ` table as described in this issue. I hope this issue wil be resolved soon.

**Comment by stefandoorn** [comment]
@magento-engcom-team Any update on this issue? We are experiencing similar problems on a database with similar amount of products. We see exactly the same happening, with the indexers running forever.

**Comment by hnsr** [comment]
This issue is fixed in 2.3.3, see https://github.com/magento/magento2/commit/995f4cbf2ae6831d9a9ae8c15d4ffe15f0183cc2#diff-3ca4f2337ba3276ff18729fdf70f63c9

**Comment by marcejohnson** [comment]
We are seeing this issue in Magento `2.3.6` for a site with 10,000+ product inventory and millions of entries in `catalogrule_product_price`.  

The **substantive** impact of this is that sale prices regress during the long reindex, causing customers to get charged the **wrong** prices if they checkout at that time.  Very bad.  And if products are changing a lot, this reindex/rebuild happens again and again.

**Details**:  After products are modified or saved, we trace the invocation of `cleanProductIndex` and `cleanProductPriceIndex` (in `/vendor/magento/module-catalog-rule/Model/Indexer/IndexBuilder.php`).  These functions perform deletes which result in table `catalogrule_product_price` getting reset/cleared as described above in the ticket description. Then all the sale prices regress. The newer commits kindly mentioned by @hnsr (which add reindexing) do not help, since the reindex still takes minutes.  

**Analysis**:  The wholesale delete in `cleanProductIndex` and `cleanProductPriceIndex` may not be a good idea.  Instead, the delete and rebuild could be performed transactionally and idempotently -- e.g. use **tmp** tables to perform the delete/rebuild first with instant swap like is done for some other reindexing.  

As a workaround to keep our rule-based sale prices stable, we apply this diff:
```
diff --git a/./vendor/magento/module-catalog-rule/Model/Indexer/IndexBuilder.php.orig b/./vendor/magento/module-catalog-rule/Model/Indexer/IndexBuilder.php
index df167d171e..935b94283e 100644
--- a/./vendor/magento/module-catalog-rule/Model/Indexer/IndexBuilder.php.orig
+++ b/./vendor/magento/module-catalog-rule/Model/Indexer/IndexBuilder.php
@@ -355,8 +355,8 @@ class IndexBuilder
      */
     private function cleanProductIndex(array $productIds): void
     {
-        $where = ['product_id IN (?)' => $productIds];
-        $this->connection->delete($this->getTable('catalogrule_product'), $where);
+        // $where = ['product_id IN (?)' => $productIds];
+        // $this->connection->delete($this->getTable('catalogrule_product'), $where);
     }
 
     /**
@@ -367,8 +367,8 @@ class IndexBuilder
      */
     private function cleanProductPriceIndex(array $productIds): void
     {
-        $where = ['product_id IN (?)' => $productIds];
-        $this->connection->delete($this->getTable('catalogrule_product_price'), $where);
+        // $where = ['product_id IN (?)' => $productIds];
+        // $this->connection->delete($this->getTable('catalogrule_product_price'), $where);
     }
 
     /**
```
As can be seen, this simply comments out the `deletes`.  We are creating a custom module / override to handle this longer term.

We see no negative effect from this diff running for weeks during high traffic/sales.
These `cleanProductIndex` and `cleanProductPriceIndex` functions were added after 2.2.x I believe... are they even necessary?
